File: training/hamic_algo/dp/787.cpp
#include <bits/stdc++.h>
using namespace std;

vector<int> a;

void solve() {
    int n = a.size();
    pair<int, int> dp[n];
    int ans = -1e9;
    for(int i = 0; i < n; ++i) {
        if (!i) dp[i].first = dp[i].second = a[i];
        else {
            if (a[i] >= 0) {
                dp[i].first = max(a[i], a[i]*dp[i-1].first);
                dp[i].second = min(a[i], a[i]*dp[i-1].second);
            } else {
                dp[i].first = max(a[i], a[i]*dp[i-1].second);
                dp[i].second = min(a[i], a[i]*dp[i-1].first);
            }
        }
        ans = max(ans, dp[i].first);
    }
    cout << ans << "\n";
}

int main() {
    int tmp;
    while(cin >> tmp) {
        if (tmp != -999999) {
            a.push_back(tmp);
        } else {
            solve();
            a.clear();
        }
    }
}

File: hamic_algo/dp/basic/787.py
import sys

def solve(a):
    n = len(a)
    if (n == 0):
        return
    dp = [[0]*2 for i in range(n)]
    ans = a[0]
    for i in range(n):
        if not i:
            dp[i][0] = a[i]
            dp[i][1] = a[i]
        else:
            if a[i] >= 0:
                dp[i][0] = max(a[i], dp[i-1][0]*a[i])
                dp[i][1] = min(a[i], dp[i-1][1]*a[i])
            else:
                dp[i][0] = max(a[i], dp[i-1][1]*a[i])
                dp[i][1] = min(a[i], dp[i-1][0]*a[i])
        ans = max(ans, dp[i][0])
    print(ans)

for line in sys.stdin:
    # print(line)
    a = list(map(int, line.split()))
    a = a[:-1]
    solve(a)
